text/0000-tracked-properties.md

```diff
@@ -0,0 +1,1029 @@
+  produced by executing a function. That value is cached until one of computed
+  property's dependencies changes.
+- A **tracked property** refers to any property that has been instrumented with
+  `@tracked`, either a _tracked getter_ or a _tracked simple property_.
+- A **tracked getter** is a JavaScript getter that has been wrapped using the
+  tracked decorator
+- A **tracked simple property** is a regular, non-getter property that has been
+  wrapped using the tracked decorator.
+- The **classic programming model** refers to the traditional Ember programming
+  model. It includes _classic classes_, _computed properties_, _event listeners_,
+  _observers_, _property notifications_, and _classic components_, and more
+  generally refers to features that will not be central to Ember Octane.
+- **Native classes** are classes defined using the Javascript `class` keyword.
+- **Classic classes** are classes defined by subclassing from `EmberObject` using
+  the static `extend` method.
+
+## Motivation
+
+Tracked properties are designed to be simpler to learn, simpler to write, and
+simpler to maintain than today's computed properties. In addition to clearer
+code, tracked properties eliminate the most common sources of bugs and mental
+model confusion in computed properties today.
+
+### Leverage Existing JavaScript Knowledge
+
+Ember's computed properties provide functionality that overlaps with native
+JavaScript getters and setters. Because native getters don't provide Ember with
+the information it needs to track changes, it's not possible to use them
+reliably in templates or in other computed properties.
+
+New learners have to "unlearn" native getters, replacing them with Ember's
+computed property system. Unfortunately, this knowledge is not portable to other
+applications that don't use Ember that developers may work on in the future, and
+while this problem may be lessened by adopting native classes and decorators, it
+still requires users learn Ember's notification system and its quirks.
+
+Tracked properties are as thin a layer as possible on top of native JavaScript.
+Tracked properties look like normal properties because they _are_ normal
+properties.
+
+Because there is no special syntax for retrieving a tracked property, any
+JavaScript syntax that feels like it should work does work:
+
+```js
+// Dot notation
+const fullName = person.fullName;
+// Destructuring
+const { fullName } = person;
+// Bracket notation for computed property names
+const fullName = person['fullName'];
+```
+
+Similarly, syntax for changing properties works just as well:
+
+```js
+// Simple assignment
+this.firstName = 'Yehuda';
+// Addition assignment (+=)
+this.lastName += 'Katz';
+// Increment operator
+this.age++;
+```
+
+This compares favorably with APIs from other libraries, which becomes more
+verbose than necessary when JavaScript syntax isn't available:
+
+```js
+this.setState({
```

Comment:
While it's nice to see how `@tracked` compares to React's `setState`, would it be more relevant for this snippet to show the current Ember `set` syntax?

```js
set(this, 'age', this.age + 1);
set(this, 'lastName', this.lastName + 'Katz');
```

Admittedly, this is pretty nit picky. Nice to see this RFC finally making the rounds. 😀

Reply:
I was assuming that most RFC readers would already be familiar with the Ember syntax. I had preemptively added this section in anticipation of any possible "Y u no just use React" comments.

Reply:
Fair enough. 😆